---
title: "R Notebook"
output: html_notebook
---

# Introduction
Scientific Question: What differences in the sequences of each segment of the segmented genome and expression of infected cells exist between the H1N1 from the 2009 Swine flu pandemic and non-pandemic H1N1 (specifically the strain A/Puerto Rico/8/1934) that may contribute to differences in transmission?

This project will be looking at the differences between the genes in the 2009 H1N1 pandemic strain (primarily A/California/04/2009) and non-pandemic strains such as A/PR/8/34 (the two strains will be referred to as CA4 and PR8). H1N1 is an influenza A virus (IAV) of the orthomyxoviridae family. It is a segmented dsDNA virus with 8 segments, each coding for a different protein. Although segments generaly code for 1 protein each, some can code for different proteins using alternative splicing and alternative start codons. Different strains of IAVs are categorized based on their hemagglutinin and neuraminidase proteins (segments 4 and 6 respectively). In addition to looking at the different sequences of segments, this project will also be looking at the differences in host cell responses by looking at RNAseq data of infected cells' transcriptome.

If there is a difference in gene sequences, then transmission related differences will be seen in the hemagglutinin and neuraminidase genes with pandemic strains having more closely correlated scores and seasonal strains having closely correlated scores, and if there is a difference in transcription of infected cells, then we should see different levels of expression for immune system related genes with pandemic strains inducing higher levels of expression of those genes.

# Background


# Data
Data for segment sequences was taken from NCBI found via fludb.org. The data for the transcription was from https://www.frontiersin.org/articles/10.3389/fmicb.2019.02344/full. Sequence data is analyzed through pairwise sequence alignment for each segment between various pandemic and seasonal strains. The scores are stored in a matrix and portrayed as a heatmap. Transcription data is in the form of the results of differential expression analysis of RPKM data. The data is analyzed through a volcano plot and pathway analysis (two analysis to make up for the lack of RNAseq due to dataset issues; as allowed by Dr. Schiffer).

---

```{r}
# Importing all necessary libraries

# The seqinr library is used to read in and perform data analysis on sequence data.Un-comment the line below to install package
# install.packages("seqinr")
library(seqinr)

# The Biostrings library is used as efficient string containers, string matching algorithm (sequence alignments), and others. Un-comment the 3 lines below to install.
# if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
# BiocManager::install("Biostrings")
library(BiocManager)
library(Biostrings)

# ggplot2 is included in R so it just needs to be loaded in. It is a library for creating plots. Here I use it to create the volcano plot.
library(ggplot2)

# The pheatmap library contains the function pheatmap() and is used to create heatmaps. Un-comment the next line to install the package.
# install.packages("pheatmap")
library(pheatmap)

# The dplyr library adds in various functions that help you manipulate data. Notably, I use the select() function for my code. Un-comment the line below to install the pacakge.
# install.packages("dplyr")
library(dplyr)

# The readxl library contains the read_excel() function I use to read in the RNAseq data. Un-comment the line below to isntall the package.
# install.packages("readxl")
library(readxl)

# The gageData library contains the data set for human biological pathways that we will be comparing our data set to. The gage library contains the gage() function that we will be using to perform the pathway analysis. The pathview library is used to generate the image of the given pathway with the relevant up and downregulated genes. Un-comment the line below to install the 3 packages.
# BiocManager::install( c("pathview", "gage", "gageData") )
library(gageData)
library(gage)
library(pathview)

# The png library will be used to read in the png files generated by pathview() as a bitmap. Un-comment the line below to install the package.
# install.packages("png")
library(png)

# The rasterImage library allows us to convert the bitmap read in by readPNG() into an image we can add to a plot. Un-comment the line below to install the package.
# install.packages("rasterImage")
library(rasterImage)
```

---

## Sequence alignment of various pandemic and seasonal strains of H1N1 and H3N2

```{r}
# loading in sequence files.Each contains 10 sequences. There are 8 segments with two segments coding for 2 proteins iether through alternative start codons or alternative splicing (m1 and m2; ns1 and ns2).

# H1N1 pandemic strain 
h1n1_ca2009 <- read.fasta("A_California_04_2009.fasta")

# H1N1 pandemic strain
h1n1_ussr1977 <-read.fasta("A_USSR_90_1977.fasta")

# H1N1 seasonal strain
h1n1_al2020 <- read.fasta("A_Alabama_01_2020.fasta")

# H1N1 seasonal strain
h1n1_pr1934 <- read.fasta("A_PR_8_1934.fasta")

# H3N2 pandemic strain
h3n2_hk1968 <- read.fasta("A_HK_1-1-MA-12_1968.fasta")

# H3N2 seasonal strain
h3n2_de2020 <- read.fasta("A_Delaware_05_2020.fasta")

# Fit all the sequence 
all_strains <- list(h1n1_ca2009, h1n1_ussr1977, h1n1_al2020, h1n1_pr1934, h3n2_hk1968, h3n2_de2020)

# Create a new empty list to store the sequences by segment instead of strain. The two numeric variables define the number of samples and number of proteins.
all_segs<-list()
num_samples<-6
num_prot <- 10

# Populating the list with all the segments.
for(i in 1:num_samples){
  all_segs$pb2[i] <- all_strains[[i]][1]
  all_segs$pb1[i] <- all_strains[[i]][2]
  all_segs$pa[i] <- all_strains[[i]][3]
  all_segs$ha[i] <- all_strains[[i]][4]
  all_segs$np[i] <- all_strains[[i]][5]
  all_segs$na[i] <- all_strains[[i]][6]
  all_segs$m1[i] <- all_strains[[i]][7]
  all_segs$m2[i] <- all_strains[[i]][8]
  all_segs$ns1[i] <- all_strains[[i]][9]
  all_segs$ns2[i] <- all_strains[[i]][10]
}

# Generate two vectors, one for all the protein names, and one for all the sample names for the heatmap.
protein_names <- c("PB2", "PB1", "PA", "HA", "NP", "NA", "M1", "M2", "NS1", "NS2")
labels <- c("h1n1_P_ca2009", "h1n1_P_ussr1977", "h1n1_S_al2020", "h1n1_S_pr1934", "h3n2_P_hk1968", "h3n2_S_de2020")

# Naming all the samples
for(i in 1:num_prot){
  names(all_segs[[i]]) <- labels
}
```

## Function for generating a heatmap based on sequence scores by taking in sequence data (as SeqFastadna data type) and the name of the protein being observed.
```{r}
heatmapMatrix <- function(sequences, protein_name) {
  scores <- matrix(nrow = length(sequences), ncol = length(sequences))
  
  # Here is a loop that populates each matrix with the appropriate score comparisons.The pairwiseAlignment function takes in two sequences in the form of a string/character vector of equal length and the score function extracts the scores. As there can be variations between the lengths of the sequences, I made all of them into a single string.
  for (i in 1:num_samples){
    
    prot1 <- paste(as.character(sequences[[i]]),collapse='')
    
    for (j in 1:num_samples){
      prot2 <- paste(as.character(sequences[[j]]),collapse='')
      scores[i,j] <- score(pairwiseAlignment(prot1, prot2))
    }
  }
  heatmap(scores, main = protein_name, labRow = labels, labCol = labels, scale = "row", cexRow = 0.65, cexCol = 0.65)
}
```


```{r}
# Here I call the function for all of the proteins that can be made by the 8 segments of influenza A.
for (i in 1:num_prot){
  heatmapMatrix(all_segs[[i]], protein_names[i])
}
```

## Analysis of sequence alignment results
Overall, the H1N1 seasonal strain from Alabama and pandemic strain from California correlate with each other most likely due to geography. Darker red indicates higher degree of similarity with the corresponding strains. The dendrograms are there as a quick way to see how each strain relates to one another in the context of the specified protein sequence.As expected, the H1N1 and H3N2 strains correlate heavily within their own groups for the HA and NA proteins as by definition the H and N refers to those two segments/proteins.Data here is more useful in tracing where segments were conserved and where they were taken from another strain within the same HxNy group. For example, both ca2009 and al2020 shares some correlation with the pr1934 seasonal strain. Interestingly enough, the pandemic strain of H1N1 in the USSR in 1977 shares a strong correlation with the pandemic strain of H3N2 in Hong Kong in 1968 for the NS1/NS2 segment/proteins. However, it is not correlated with the pandemic strain in California in 2009. Co-infection of a host by a descendant of the 1968 strain and a predecessor of the 1977 strain could have allowed exchange of the segment. It could also be a case of convergent evolution towards a more similar NS1/NS2. As the non-structural proteins are known to be used to aid in repressing host interferon pathway and pro-inflammatory responses (Clark 2017). It would further benefit to see the sequential and structural differences and similarities of various strains' NS1 and NS2 genes and proteins, respectively.

---

# RNAseq analysis of RPKM

In this section I will be reading in finished RNAseq data (post-DEseq) and be analyzing the data using heatmap, volacano plot, and pathway analysis. Each analysis method will have its own superficial description/analysis at the end of their section, and there will be an overall analysis discussing the significance of the findings and the functions of previously mentioned genes and pathways.

## Read in data and transform data to fit further analysis
```{r}
# Here I read in the data from the provided files that include the log2ratio of infected/control expressions.
res_ca01 <- as.data.frame(read_excel("A549_CA01.xlsx", col_names = TRUE))
res_pr8 <- as.data.frame(read_excel("A549_PR8.xlsx", col_names = TRUE))

# After reading in the table, I do some transformations so I end up with only the log2ratios and pvalues.
rownames(res_ca01) <- res_ca01[,11]
ca01_geneid <- res_ca01[,1]
res_ca01 <- res_ca01[,-8]
res_ca01 <- res_ca01[,7:8]
res_ca01$geneID <- ca01_geneid
res_ca01 <- res_ca01[order(res_ca01$pvalue),]
rownames(res_pr8) <- res_pr8[,11]
pr8_geneid <- res_pr8[,1]
res_pr8 <- res_pr8[,-8]
res_pr8 <- res_pr8[,7:8]
res_pr8$geneID <- pr8_geneid
res_pr8 <- res_pr8[order(res_pr8$pvalue),]

# I subset it to only include the statistically significant. Although it may be more beneficial to look at all the genes in the seasonal strain. It can also be interesting to look to see if upregulated genes in ca01 infected cells are downregulated in PR8 infected cells.
res_ca01_sig <- subset(res_ca01, res_ca01$pvalue < 0.05)
res_pr8_sig <- subset(res_pr8, res_ca01$pvalue < 0.05)
```

```{r}
# To make analysis simpler, I subseted both to only include genes that overlap between the two.
sub_res_pr8 <- subset(res_pr8_sig, rownames(res_pr8_sig) %in% rownames(res_ca01_sig))
sub_res_ca01 <- subset(res_ca01_sig, rownames(res_ca01_sig) %in% rownames(sub_res_pr8))

# I combine the two data frames to draw the heatmap and volcano plot for later.
combined_res <- data.frame(sub_res_ca01$`log2_Ratio(A549_CA01/A549_Control)`, sub_res_pr8$`log2_Ratio(A549_PR8/A549_Control)`, sub_res_ca01$pvalue, sub_res_pr8$pvalue)
rownames(combined_res)<-rownames(sub_res_pr8)
colnames(combined_res) <- c("CA01_infected", "PR8_infected", "CA01_pvalue", "PR8_pvalue")
```


## Generating heatmap
```{r}
# I generate a heatmap for the top 10 DEGs.
sub_combined <- head(as.matrix(combined_res), n =10)
pheatmap(sub_combined[,1:2],
         cluster_rows = T, 
         show_rownames = T, row.names = rownames(sub_combined[,1:2]), treeheight_row = 0, treeheight_col = 0)
```

```{r}
# In this chunk, I am interested in seeing if there are up regulated genes in the CA01 group that are down regulated in the PR8 group and vice versa.
res_ca01_up <- subset(res_ca01_sig, res_ca01_sig$`log2_Ratio(A549_CA01/A549_Control)` > 0)
res_ca01_down <- subset(res_ca01_sig, res_ca01_sig$`log2_Ratio(A549_CA01/A549_Control)` < 0)
res_pr8_up <- subset(res_pr8_sig, res_pr8_sig$`log2_Ratio(A549_PR8/A549_Control)` > 0)
res_pr8_down <- subset(res_pr8_sig, res_pr8_sig$`log2_Ratio(A549_PR8/A549_Control)` < 0)

ca01_up_pr8_down <- subset(res_ca01_up, rownames(res_pr8_down) %in% rownames(res_ca01_up))
ca01_down_pr8_up <- subset(res_ca01_down, rownames(res_pr8_up) %in% rownames(res_ca01_down))

# Prints out genes that are upregulated in CA01 infected A549 cells and are downregulated in PR8 infected A549 cells.
rownames(ca01_up_pr8_down)

# Prints out genes that are downregulated in CA01 infected A549 cells and are upregulated in PR8 infected A549 cells.
rownames(ca01_down_pr8_up)
```
### Analysis of heatmap
  Here are the top 10 DEGs of CA01 infected cells along with the corresponding levels of expression in PR8 infected cells. Level of expression here is represented by the log2 ratio of infected RPKM/control RPKM. Of note are IFIT1 and PARP10. IFIT1 is the highest DEG in CA01 infected cells. OASL seems to be the DEG with the greatest difference in expression between the CA01 and PR8 infected A549 cells. I will go more in depth into IFIT1 and OASL in the overall analysis at the end.

## Generating volcano plot
```{r}
# Here I generated the volcano plot.
combined_res$CA01_threshold <- combined_res$CA01_pvalue < 0.05
combined_res$PR8_threshold <- combined_res$PR8_pvalue < 0.05
ggplot(combined_res) + 
        geom_point(aes(x = CA01_infected, y = -log10(CA01_pvalue), color = CA01_threshold)) + 
        xlab("log2_Ratio(A549_CA01/A549_Control)") + 
        ylab("-log10 p-value") + 
        theme(legend.position = "none", 
              plot.title = element_text(size = rel(1.5), hjust = 0.5), 
              axis.title = element_text(size = rel(1.25)))

ggplot(combined_res) + 
        geom_point(aes(x = PR8_infected, y = -log10(PR8_pvalue), color = PR8_threshold)) + 
        xlab("log2_Ratio(A549_PR8/A549_Control)") + 
        ylab("-log10 p-value") + 
        theme(legend.position = "none", 
              plot.title = element_text(size = rel(1.5), hjust = 0.5), 
              axis.title = element_text(size = rel(1.25)))
```
### Analysis of volcano plots
  Here in red are all the DEGs with p-value < 0.05. DEGs with p-value > 0.05 are shown in blue. As you can see all DEGs here are statistically significant. However, upon further examination, the dataset from the paper only includes statistically significant DEGs. However, analysis is still possible. Overall the main difference is that CA01 infected cells have one DEG that is much more downregulated. The line below shows the DEG, which is TNFAIP8L3.
```{r}
combined_res[combined_res$CA01_infected < -5,]
```


## Pathway analysis of significant results

```{r}
# Load in dataset for human genes
data(kegg.sets.hs)

# The gage function takes in a vector of foldchanges with the name of each element being the geneid. It returns a list of pathways with the increased and decreased pathways stored in "greater" and "less" respectively.
foldchanges_ca01 <- res_ca01$`log2_Ratio(A549_CA01/A549_Control)`
names(foldchanges_ca01) <- res_ca01$geneID
keggres_ca01 <- gage(foldchanges_ca01, gsets=kegg.sets.hs)

foldchanges_pr8 <- res_pr8$`log2_Ratio(A549_PR8/A549_Control)`
names(foldchanges_pr8) <- res_pr8$geneID
keggres_pr8 <- gage(foldchanges_pr8, gsets=kegg.sets.hs)

#removing all NA values
keggres_ca01$greater <- keggres_ca01$greater[!is.na(keggres_ca01$greater[,1]),]
keggres_ca01$less <- keggres_ca01$less[!is.na(keggres_ca01$less[,1]),]
keggres_ca01$stats <- keggres_ca01$stats[!is.na(keggres_ca01$stats[,1]),]

#removing all NA values

keggres_pr8$greater <- keggres_pr8$greater[!is.na(keggres_pr8$greater[,1]),]
keggres_pr8$less <- keggres_pr8$less[!is.na(keggres_pr8$less[,1]),]
keggres_pr8$stats <- keggres_pr8$stats[!is.na(keggres_pr8$stats[,1]),]

#Subsetting significant pathways (p-value < 0.05)
ca01_up_path <- keggres_ca01$greater[keggres_ca01$greater[,3] < 0.05,]
ca01_down_path <- keggres_ca01$less[keggres_ca01$less[,3] < 0.05,]
pr8_up_path <- keggres_pr8$greater[keggres_pr8$greater[,3] < 0.05,]
pr8_down_path <- keggres_pr8$less[keggres_pr8$less[,3] < 0.05,]

# Additional naming needed for ca01_down_path is needed because there is only 1 pathway with p-value < 0.05, leading to a vector being returned instead of a matrix.
ca01_down_path <- matrix(ca01_down_path, nrow = 1, ncol = 6)
rownames(ca01_down_path) <- rownames(keggres_ca01$less)[1]
colnames(ca01_down_path) <- colnames(keggres_ca01$less)

# Combining significant pathways into one list.
pathway_results <- list(ca01_up_path = ca01_up_path, ca01_down_path = ca01_down_path, pr8_up_path = pr8_up_path, pr8_down_path = pr8_down_path)
```

```{r}
# Print out pathways up and down regulated with p-values < 0.05 for infection of A549 cells with ca01 and PR8 strains of H1N1
writeLines("Upregualted pathways in CA01 infected A549 cells")
ca01_up_path

writeLines("\n\nUpregualted pathways in PR8 infected A549 cells")
pr8_up_path

writeLines("\n\nDownregualted pathways in CA01 infected A549 cells")
ca01_down_path

writeLines("\n\nDownregualted pathways in PR8 infected A549 cells")
pr8_down_path
```
```{r}
# Funciton I made to more easily generate the pathway analysis image and load in the generated png. Due to the size and resolution of the images, they are not vieable in the html file. However, it is viewable in R studio by clicking on "Show in new window" after the images are generated.
displayPathView <- function(foldchanges, pathway, group = ""){
  pathview(gene.data = foldchanges, pathway.id = pathway, out.suffix = group)
  img <- readPNG(paste(pathway, group, "png", sep = "."))
  plot.new()
  title(paste(group, pathway, sep = ": "))
  rasterImage(img,0,0,1,1)
}
```


```{r}
# Here I look at the hsa05160 pathway, which is the Hepatitis C infection pathway, as it is upregulated in both types of infections.
displayPathView(foldchanges_ca01, "hsa05160", "ca01_up")
displayPathView(foldchanges_pr8, "hsa05160", "pr8_up")

# Here I look at the hsa04620 pathway, which is the Toll-like receptor pathway, as it is upregulated only in CA01 infected A549 cells.
displayPathView(foldchanges_ca01, "hsa04620", "ca01_up")

# Here I look at the hsa04510 pathway, which is the focal adhesion pathway, as it is downregulated in both types of infections.
displayPathView(foldchanges_ca01, "hsa04510", "ca01_down")

# Here I look at the hsa04512 pathway, which is the ECM-receptor interaction pathway, as it is downregulated only in PR8 infected cells.
displayPathView(foldchanges_pr8, "hsa04512", "pr8_down")

# Here I look at the hsa05200 pathway, which is Pathways in cancer, as it is downregulated only in PR8 infected cells.
displayPathView(foldchanges_pr8, "hsa05200", "pr8_down")

```
### Analysis of pathway analysis results
We see that CA01 infected cells uniquely showed up-regulation in Toll-like receptor signaling, RIG-I-like receptor signaling, and cystolic DNA-sensing pathway, in addition to sharing with PR8 infected cells the up-regulated pathways of Hepatitis C infection pathway and JAK-STAT signaling pathway. We also see that PR8 infected cells uniquely showed down-regulation in the ECM-receptor interaction pathway and pathways in cancer, in addition to sharing with CA01 infected cells the down-regulated pathway of focal adhesion.

## Overall analysis/discussion/conclusion of RNAseq data based on RPKM

IFIT1 stands for interferon-induced protein with tetratricopeptide repeats 1 and is a protein reported to have cell-intrinsic antiviral activity against negative sense ssRNA viruses, such as H1N1 (Pinto 2015). It goes to reason that a more serious infection (CA01 pandemic as opposed to PR8 seasonal) will lead to higher expression of IFIT1.

OASL stands for oligoadenylate synthetases-like protein and is induced by interferons. OASL is also mediated by the RIG-I pathway, which is a viral RNA sensor pathway (Zhu 2014). Like mentioned earlier, OASL is the most different DEG between CA01 and PR8 infected cells. Without further examination, there can be no way of determining the reason. However, one possibility could be that the 2009 pandemic strain of viruses lead to much a much higher number of viral progeny. This can lead to more copies of the genome being detected in the original host cell and more viruses infecting a single cell. Another possibility could be that the pandemic strain has an earlier and higher peak in transcription of positive strength RNA, meaning it is faster at infecting the cell and at manipulating host machinery.

In terms of pathway analysis it is more useful to look at pathways that are only up or down regulated in one group.
Upregulated in only CA01: TLR signaling pathway, RIG-I-like signaling pathway, and cytosolic DNA-sensing pathway.
Downregulated in only PR8: ECM-receptor interaction pathway and pathways in cancer.

# References

Clark, A. M., Nogales, A., Martinez-Sobrido, L., Topham, D. J., & DeDiego, M. L. (2017). Functional Evolution of Influenza Virus NS1 Protein in Currently Circulating Human 2009 Pandemic H1N1 Viruses. Journal of virology, 91(17), e00721-17. https://doi.org/10.1128/JVI.00721-17

Pinto, A. K., Williams, G. D., Szretter, K. J., White, J. P., Proença-Módena, J. L., Liu, G., Olejnik, J., Brien, J. D., Ebihara, H., Mühlberger, E., Amarasinghe, G., Diamond, M. S., &amp; Boon, A. C. (2015). Human and murine IFIT1 proteins do not restrict infection of negative-sense RNA viruses of the Orthomyxoviridae, Bunyaviridae, and Filoviridae families. Journal of Virology, 89(18), 9465–9476. https://doi.org/10.1128/jvi.00996-15 

Zhu, J., Zhang, Y., Ghosh, A., Cuevas, R. A., Forero, A., Dhar, J., Ibsen, M. S., Schmid-Burgk, J. L., Schmidt, T., Ganapathiraju, M. K., Fujita, T., Hartmann, R., Barik, S., Hornung, V., Coyne, C. B., & Sarkar, S. N. (2014). Antiviral activity of human OASL protein is mediated by enhancing signaling of the RIG-I RNA sensor. Immunity, 40(6), 936–948. https://doi.org/10.1016/j.immuni.2014.05.007